// enzyme-adapter-preact-pure/build-cjs/src/StringRenderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var preact_1 = require("preact");
var StringRenderer = /** @class */ (function () {
    function StringRenderer() {
    }
    StringRenderer.prototype.render = function (el, context) {
        var tempContainer = document.createElement('div');
        preact_1.render(el, tempContainer);
        var html = tempContainer.innerHTML;
        preact_1.render(preact_1.h('unmount-me', {}), tempContainer);
        return html;
    };
    StringRenderer.prototype.simulateError = function (nodeHierarchy, rootNode, error) {
        throw new Error('Static rendering does not support simulating errors');
    };
    StringRenderer.prototype.simulateEvent = function (node, eventName, args) {
        throw new Error('Static rendering does not support simulating events');
    };
    StringRenderer.prototype.unmount = function () {
        // No-op
    };
    StringRenderer.prototype.getNode = function () {
        return null;
    };
    StringRenderer.prototype.batchedUpdates = function (fn) {
        fn();
    };
    return StringRenderer;
}());
exports.default = StringRenderer;
