// iris-lib/cjs/components/FollowButton.js
'use strict';

exports.__esModule = true;

var _templateObject = _taggedTemplateLiteralLoose(['\n      <button class="iris-follow-button ', ' ', '" onClick=', '>\n        <span class="nonhover">', '</span>\n        <span class="hover">Unfollow</span>\n      </button>\n    '], ['\n      <button class="iris-follow-button ', ' ', '" onClick=', '>\n        <span class="nonhover">', '</span>\n        <span class="hover">Unfollow</span>\n      </button>\n    ']);

var _preact = require('preact');

var _util = require('../util');

var _util2 = _interopRequireDefault(_util);

var _preact2 = require('htm/preact');

var _Key = require('../Key');

var _Key2 = _interopRequireDefault(_Key);

var _preactCustomElement = require('preact-custom-element');

var _preactCustomElement2 = _interopRequireDefault(_preactCustomElement);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FollowButton = function (_Component) {
  _inherits(FollowButton, _Component);

  function FollowButton() {
    _classCallCheck(this, FollowButton);

    var _this = _possibleConstructorReturn(this, _Component.call(this));

    _this.eventListeners = {};
    return _this;
  }

  FollowButton.prototype.onClick = function onClick(e) {
    e.preventDefault();
    var follow = !this.state.following;
    _util2.default.getPublicState().user().get('follow').get(this.props.user).put(follow);
  };

  FollowButton.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    _util2.default.injectCss();
    _Key2.default.getDefault().then(function (key) {
      _util2.default.getPublicState().user().auth(key);
      _util2.default.getPublicState().user().get('follow').get(_this2.props.user).on(function (following, a, b, e) {
        _this2.setState({ following: following });
        _this2.eventListeners['follow'] = e;
      });
    });
  };

  FollowButton.prototype.componentWillUnmount = function componentWillUnmount() {
    Object.values(this.eventListeners).forEach(function (e) {
      return e.off();
    });
  };

  FollowButton.prototype.render = function render() {
    var _this3 = this;

    return (0, _preact2.html)(_templateObject, this.state.following ? 'following' : '', this.props['inner-class'] || '', function (e) {
      return _this3.onClick(e);
    }, this.state.following ? 'Following' : 'Follow');
  };

  return FollowButton;
}(_preact.Component);

!_util2.default.isNode && (0, _preactCustomElement2.default)(FollowButton, 'iris-follow-button', ['user']);

exports.default = FollowButton;
module.exports = exports['default'];
