// iris-lib/cjs/index.js
'use strict';

exports.__esModule = true;

var _Collection = require('./Collection');

var _Collection2 = _interopRequireDefault(_Collection);

var _SignedMessage = require('./SignedMessage');

var _SignedMessage2 = _interopRequireDefault(_SignedMessage);

var _Contact = require('./Contact');

var _Contact2 = _interopRequireDefault(_Contact);

var _Attribute = require('./Attribute');

var _Attribute2 = _interopRequireDefault(_Attribute);

var _util = require('./util');

var _util2 = _interopRequireDefault(_util);

var _Key = require('./Key');

var _Key2 = _interopRequireDefault(_Key);

var _Channel = require('./Channel');

var _Channel2 = _interopRequireDefault(_Channel);

var _package = require('../package.json');

var _Identicon = require('./components/Identicon');

var _Identicon2 = _interopRequireDefault(_Identicon);

var _TextNode = require('./components/TextNode');

var _TextNode2 = _interopRequireDefault(_TextNode);

var _ImageNode = require('./components/ImageNode');

var _ImageNode2 = _interopRequireDefault(_ImageNode);

var _CopyButton = require('./components/CopyButton');

var _CopyButton2 = _interopRequireDefault(_CopyButton);

var _FollowButton = require('./components/FollowButton');

var _FollowButton2 = _interopRequireDefault(_FollowButton);

var _Search = require('./components/Search');

var _Search2 = _interopRequireDefault(_Search);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*eslint no-useless-escape: "off", camelcase: "off" */

exports.default = {
  VERSION: _package.version,
  Collection: _Collection2.default,
  SignedMessage: _SignedMessage2.default,
  Contact: _Contact2.default,
  Attribute: _Attribute2.default,
  Key: _Key2.default,
  Channel: _Channel2.default,
  util: _util2.default,
  components: {
    Identicon: _Identicon2.default,
    TextNode: _TextNode2.default,
    ImageNode: _ImageNode2.default,
    CopyButton: _CopyButton2.default,
    FollowButton: _FollowButton2.default,
    Search: _Search2.default
  }
};
module.exports = exports['default'];
