// @strata-foundation/react/dist/lib/contexts/strataProgramsContext.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StrataProgramsProvider = exports.StrataProgramsContext = void 0;
const react_async_hook_1 = require("react-async-hook");
const useProvider_1 = require("../hooks/useProvider");
const spl_token_bonding_1 = require("@strata-foundation/spl-token-bonding");
const spl_token_collective_1 = require("@strata-foundation/spl-token-collective");
const react_1 = __importDefault(require("react"));
exports.StrataProgramsContext = react_1.default.createContext({
    loading: true
});
;
function getPrograms(provider) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!provider) {
            return {};
        }
        const tokenCollective = yield spl_token_collective_1.SplTokenCollective.init(provider);
        const tokenBonding = yield spl_token_bonding_1.SplTokenBonding.init(provider);
        return {
            tokenCollectiveSdk: tokenCollective,
            tokenBondingSdk: tokenBonding,
        };
    });
}
const StrataProgramsProvider = ({ children }) => {
    const provider = (0, useProvider_1.useProvider)();
    const { result, loading, error } = (0, react_async_hook_1.useAsync)(getPrograms, [provider]);
    return react_1.default.createElement(exports.StrataProgramsContext.Provider, { value: {
            tokenCollectiveSdk: result === null || result === void 0 ? void 0 : result.tokenCollectiveSdk,
            tokenBondingSdk: result === null || result === void 0 ? void 0 : result.tokenBondingSdk,
            error,
            loading
        } }, children);
};
exports.StrataProgramsProvider = StrataProgramsProvider;
