// iris-lib/es/components/FollowButton.js
var _templateObject = _taggedTemplateLiteralLoose(['\n      <button class="iris-follow-button ', ' ', '" onClick=', '>\n        <span class="nonhover">', '</span>\n        <span class="hover">Unfollow</span>\n      </button>\n    '], ['\n      <button class="iris-follow-button ', ' ', '" onClick=', '>\n        <span class="nonhover">', '</span>\n        <span class="hover">Unfollow</span>\n      </button>\n    ']);

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import { Component } from 'preact';
import util from '../util';
import { html } from 'htm/preact';
import Key from '../Key';
import register from 'preact-custom-element';

var FollowButton = function (_Component) {
  _inherits(FollowButton, _Component);

  function FollowButton() {
    _classCallCheck(this, FollowButton);

    var _this = _possibleConstructorReturn(this, _Component.call(this));

    _this.eventListeners = {};
    return _this;
  }

  FollowButton.prototype.onClick = function onClick(e) {
    e.preventDefault();
    var follow = !this.state.following;
    util.getPublicState().user().get('follow').get(this.props.user).put(follow);
  };

  FollowButton.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    util.injectCss();
    Key.getDefault().then(function (key) {
      util.getPublicState().user().auth(key);
      util.getPublicState().user().get('follow').get(_this2.props.user).on(function (following, a, b, e) {
        _this2.setState({ following: following });
        _this2.eventListeners['follow'] = e;
      });
    });
  };

  FollowButton.prototype.componentWillUnmount = function componentWillUnmount() {
    Object.values(this.eventListeners).forEach(function (e) {
      return e.off();
    });
  };

  FollowButton.prototype.render = function render() {
    var _this3 = this;

    return html(_templateObject, this.state.following ? 'following' : '', this.props['inner-class'] || '', function (e) {
      return _this3.onClick(e);
    }, this.state.following ? 'Following' : 'Follow');
  };

  return FollowButton;
}(Component);

!util.isNode && register(FollowButton, 'iris-follow-button', ['user']);

export default FollowButton;
