// iris-lib/es/components/Search.js
var _templateObject = _taggedTemplateLiteralLoose(['\n      <div class="iris-search-box" style="position: relative;">\n        <form onSubmit=', '>\n          <label>\n            <input class="', '" type="text" placeholder="Search" onInput=', '/>\n          </label>\n        </form>\n        <', ' class="search-box-results" style="position: absolute; background-color: white; border: 1px solid #eee; border-radius: 8px; left: ', '">\n          ', '\n          ', '\n        <//>\n      </div>\n    '], ['\n      <div class="iris-search-box" style="position: relative;">\n        <form onSubmit=', '>\n          <label>\n            <input class="', '" type="text" placeholder="Search" onInput=', '/>\n          </label>\n        </form>\n        <', ' class="search-box-results" style="position: absolute; background-color: white; border: 1px solid #eee; border-radius: 8px; left: ', '">\n          ', '\n          ', '\n        <//>\n      </div>\n    ']),
    _templateObject2 = _taggedTemplateLiteralLoose(['\n              <a style="width: 300px; display: flex; padding: 5px; flex-direction: row" href="https://iris.to/profile/', '" onClick=', '>\n                <', ' user=', ' width=40/>\n                <', ' marginLeft="5px">\n                  ', '<br/>\n                  <small>\n                    ', '\n                  </small>\n                <//>\n              <//>\n            '], ['\n              <a style="width: 300px; display: flex; padding: 5px; flex-direction: row" href="https://iris.to/profile/', '" onClick=', '>\n                <', ' user=', ' width=40/>\n                <', ' marginLeft="5px">\n                  ', '<br/>\n                  <small>\n                    ', '\n                  </small>\n                <//>\n              <//>\n            ']),
    _templateObject3 = _taggedTemplateLiteralLoose(['\n            <a class="follow-someone" style="padding:5px;">Follow someone to see more search results</a>\n            <a style="width: 300px; display: flex; padding: 5px; flex-direction: row" onClick=', ' href="https://iris.to/profile/', '" class="suggested">\n              <', ' user=', ' width=40/>\n              <', ' alignItems="center" marginLeft="5px"><i>Suggested</i><//>\n            </a>\n          '], ['\n            <a class="follow-someone" style="padding:5px;">Follow someone to see more search results</a>\n            <a style="width: 300px; display: flex; padding: 5px; flex-direction: row" onClick=', ' href="https://iris.to/profile/', '" class="suggested">\n              <', ' user=', ' width=40/>\n              <', ' alignItems="center" marginLeft="5px"><i>Suggested</i><//>\n            </a>\n          ']);

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import register from 'preact-custom-element';
import { Component } from 'preact';
import { html } from 'htm/preact';
import { Row, Col } from 'jsxstyle/preact';
import util from '../util';
import Key from '../Key';
import Identicon from './Identicon.js';
import Fuse from 'fuse.js';

var suggestedFollow = 'hyECQHwSo7fgr2MVfPyakvayPeixxsaAWVtZ-vbaiSc.TXIp8MnCtrnW6n2MrYquWPcc-DTmZzMBmc2yaGv9gIU';

var Search = function (_Component) {
  _inherits(Search, _Component);

  function Search() {
    _classCallCheck(this, Search);

    var _this = _possibleConstructorReturn(this, _Component.call(this));

    _this.eventListeners = {};
    _this.state = { results: [] };
    _this.follows = {};
    _this.debouncedIndexAndSearch = util.debounce(function () {
      var options = { keys: ['name'], includeScore: true, includeMatches: true, threshold: 0.3 };
      _this.fuse = new Fuse(Object.values(_this.follows), options);
      _this.search();
    }, 200);
    Key.getDefault().then(function (key) {
      _this.key = key;
      util.getPublicState().user().auth(key);
      _this.getFollowsFn(function () {
        return _this.debouncedIndexAndSearch();
      });
    });
    return _this;
  }

  Search.prototype.onInput = function onInput() {
    this.search();
  };

  Search.prototype.close = function close() {
    this.base.querySelector('input').value = '';
    this.setState({ results: [], query: '' });
  };

  Search.prototype.getFollowsFn = function getFollowsFn(callback, k) {
    var _this2 = this;

    var maxDepth = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 2;
    var currentDepth = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 1;

    k = k || this.key.pub;

    var addFollow = function addFollow(k, followDistance, follower) {
      if (_this2.follows[k]) {
        if (_this2.follows[k].followDistance > followDistance) {
          _this2.follows[k].followDistance = followDistance;
        }
        _this2.follows[k].followers.add(follower);
      } else {
        _this2.follows[k] = { key: k, followDistance: followDistance, followers: new Set([follower]) };
        util.getPublicState().user(k).get('profile').get('name').on(function (name) {
          _this2.follows[k].name = name;
          callback(k, _this2.follows[k]);
        });
      }
      callback(k, _this2.follows[k]);
    };

    addFollow(k, currentDepth - 1);

    util.getPublicState().user(k).get('follow').map().once(function (isFollowing, followedKey) {
      // TODO: .on for unfollow
      if (isFollowing) {
        _this2.hasFollows = true;
        addFollow(followedKey, currentDepth, k);
        if (currentDepth < maxDepth) {
          _this2.getFollowsFn(callback, followedKey, maxDepth, currentDepth + 1);
        }
      }
    });

    return this.follows;
  };

  Search.prototype.componentDidMount = function componentDidMount() {
    this.adjustResultsPosition();
  };

  Search.prototype.componentDidUpdate = function componentDidUpdate() {
    this.adjustResultsPosition();
  };

  Search.prototype.adjustResultsPosition = function adjustResultsPosition() {
    var input = this.base.querySelector('input');
    this.offsetLeft = input.offsetLeft;
  };

  Search.prototype.componentWillUnmount = function componentWillUnmount() {
    Object.values(this.eventListeners).forEach(function (e) {
      return e.off();
    });
  };

  Search.prototype.onSubmit = function onSubmit(e) {
    e.preventDefault();
    var links = this.base.querySelector('a:not(.follow-someone)');
    links.length && links[0].click();
    this.base.querySelector('input').blur();
  };

  Search.prototype.search = function search() {
    var _this3 = this;

    var query = this.base.querySelector('input').value;

    if (this.props['on-select']) {
      var s = query.split('https://iris.to/profile/');
      if (s.length > 1) {
        return this.props['on-select']({ key: s[1] });
      }
      var key = null; //Helpers.getUrlParameter('chatWith', s[1]);
      if (key) {
        return this.props['on-select']({ key: key });
      }
    }
    //if (followChatLink(query)) return;

    if (query && this.fuse) {
      var results = this.fuse.search(query).slice(0, 5);
      if (results.length) {
        var onKeyUp = function onKeyUp(e) {
          if (e.key === 'Escape') {
            // escape key maps to keycode `27`
            document.removeEventListener('keyup', onKeyUp);
            _this3.close();
          }
        };
        document.removeEventListener('keyup', onKeyUp);
        document.addEventListener('keyup', onKeyUp);
      }
      this.setState({ results: results, query: query });
    } else {
      this.setState({ results: [], query: query });
    }
  };

  Search.prototype.onClick = function onClick(e, item) {
    this.close();
    var onSelect = this.props.onSelect || window.onIrisSearchSelect;
    if (onSelect) {
      e.preventDefault();
      e.stopPropagation();
      onSelect(item);
    }
  };

  Search.prototype.render = function render() {
    var _this4 = this;

    return html(_templateObject, function (e) {
      return _this4.onSubmit(e);
    }, this.props['inner-class'] || '', function () {
      return _this4.onInput();
    }, Col, this.offsetLeft || '', this.state.results.map(function (r) {
      var i = r.item;
      var followText = '';
      if (i.followDistance === 1) {
        followText = 'Following';
      }
      if (i.followDistance === 2) {
        if (i.followers.size === 1 && _this4.follows[[].concat(i.followers)[0]] && _this4.follows[[].concat(i.followers)[0]].name) {
          followText = 'Followed by ' + _this4.follows[[].concat(i.followers)[0]].name;
        } else {
          followText = 'Followed by ' + i.followers.size + ' users you follow';
        }
      }
      return html(_templateObject2, i.key, function (e) {
        return _this4.onClick(e, i);
      }, Identicon, i.key, Col, i.name || '', followText);
    }), this.state.query && !this.hasFollows ? html(_templateObject3, function (e) {
      return _this4.onClick(e, { key: suggestedFollow });
    }, suggestedFollow, Identicon, suggestedFollow, Row) : '');
  };

  return Search;
}(Component);

!util.isNode && register(Search, 'iris-search', ['on-select', 'inner-class']);

export default Search;
